pywinauto/linux/keyboard.py

```diff
@@ -0,0 +1,550 @@
+            modifiers.append(c)
+            # hold down the modifier key
+            #keys.append(KeyAction(modifier, up = False))
+            if DEBUG:
+                print("MODS+", modifiers)
+            continue
+
+        # Apply modifiers over a bunch of characters (not just one!)
+        elif c == "(":
+            # find the end of the bracketed text
+            end_pos = string.find(")", index)
+            if end_pos == -1:
+                raise KeySequenceError('`)` not found')
+            keys.extend(
+                parse_keys(string[index:end_pos], modifiers = modifiers))
+            index = end_pos + 1
+
+        # Escape or named key
+        elif c == "{":
+            # We start searching from index + 1 to account for the case {}}
+            end_pos = string.find("}", index+1)
+            if end_pos == -1:
+                raise KeySequenceError('`}` not found')
+
+            code = string[index:end_pos]
+            index = end_pos + 1
+            keys.extend(handle_code(code))
+
+        # unmatched ")"
+        elif c == ')':
+            raise KeySequenceError('`)` should be preceeded by `(`')
+
+        # unmatched "}"
+        elif c == '}':
+            raise KeySequenceError('`}` should be preceeded by `{`')
+
+        # so it is a normal character
+        else:
+            # don't output white space unless flags to output have been set
+            if (c == ' ' and not with_spaces or
+                c == '\t' and not with_tabs or
+                c == '\n' and not with_newlines):
+                continue
+
+            # output newline
+            if c in ('~', '\n'):
+                keys.append(KeyAction(CODES["ENTER"]))
+
+            elif modifiers:
+                keys.append(KeyAction(c))
+
+            else:
+                keys.append(KeyAction(c))
+
+        # as we have handled the text - release the modifiers
+        while modifiers:
+            if DEBUG:
+                print("MODS-", modifiers)
+            mod = modifiers.pop()
+            if mod == '+':
+                keys[-1].shift = True
+            elif mod == '^':
+                keys[-1].ctrl = True
+            elif mod == '%':
+                keys[-1].alt = True
+
+    # just in case there were any modifiers left pressed - release them
+    while modifiers:
+        keys.append(KeyAction(modifiers.pop(), down = False))
+    return keys
+
+
+def SendKeys(keys,
+             pause=0.05,
+             with_spaces=False,
+             with_tabs=False,
+             with_newlines=False,
+             turn_off_numlock=True):
+    """Parse the keys and type them"""
+    keys = parse_keys(keys, with_spaces, with_tabs, with_newlines)
+    for k in keys:
+        k.run()
+        time.sleep(pause)
+
+
+def main(): #pragma: no cover
+    """Send some test strings"""
+    actions = """
+        {LWIN}
+        {PAUSE .25}
+        r
+        {PAUSE .25}
+        Notepad.exe{ENTER}
```

Comment:
This obsolete old code (the whole `main()` function) could be removed from the Linux module.